File: src/main/java/net/nodus/security/auth/service/JwtTokenService.java
package net.nodus.security.auth.service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Date;
import java.util.Objects;
import javax.crypto.SecretKey;
import net.nodus.database.account.UserAccount;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class JwtTokenService {

    private final String secret;
    private final long expirationSeconds;

    public JwtTokenService(
        @Value("${jwt.secret}") String secret,
        @Value("${jwt.access-token-expiration-seconds}") long expirationSeconds
    ) {
        this.secret = secret;
        this.expirationSeconds = expirationSeconds;
    }

    public String createAccessToken(UserAccount user) {
        String userId = Objects.requireNonNull(user.getId());
        Instant now = Instant.now();
        Instant expiresAt = now.plusSeconds(expirationSeconds);

        return Jwts.builder()
            .subject(userId)
            .claim("email", user.getEmail())
            .claim("name", user.getName())
            .issuedAt(Date.from(now))
            .expiration(Date.from(expiresAt))
            .signWith(signingKey(), Jwts.SIG.HS256)
            .compact();
    }

    public Claims parseClaims(String token) {
        return Jwts.parser()
            .verifyWith(signingKey())
            .build()
            .parseSignedClaims(token)
            .getPayload();
    }

    private SecretKey signingKey() {
        return Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
    }
}

File: src/main/kotlin/net/nodus/auth/JwtTokenService.kt
package net.nodus.auth

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import net.nodus.account.UserAccount
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.Date
import javax.crypto.SecretKey

@Service
class JwtTokenService(
    @Value("\${jwt.secret}")
    private val secret: String,

    @Value("\${jwt.access-token-expiration-seconds}")
    private val expirationSeconds: Long,
) {

    private val signingKey: SecretKey
        get() = Keys.hmacShaKeyFor(secret.toByteArray())

    fun createAccessToken(user: UserAccount): String {
        val userId = requireNotNull(user.id)
        val now = Instant.now()
        val expiresAt = now.plusSeconds(expirationSeconds)

        return Jwts.builder()
            .subject(userId)
            .claim("email", user.email)
            .claim("name", user.name)
            .issuedAt(Date.from(now))
            .expiration(Date.from(expiresAt))
            .signWith(signingKey, Jwts.SIG.HS256)
            .compact()
    }

    fun parseClaims(token: String): Claims =
        Jwts.parser()
            .verifyWith(signingKey)
            .build()
            .parseSignedClaims(token)
            .payload

}
